typed-racket-lib/typed-racket/typecheck/tc-app/tc-app-list.rkt

```diff
@@ -78,21 +78,29 @@
         ;; TODO fix double typechecking
         [_ (tc/app-regular #'form expected)])))
   ;; special case for `list'
-  (pattern (list . args)
-    (let ()
+  (pattern
```

Comment:
I implemented the `(List ...)` simplification, but the union case I think is best to stick to what we're currently doing. It's not the case that we could simply try some inferences and not others, we need to get all possible substitutions and try them out as we do now.